File: 1.Two Sum.py
# Problem Link: https://leetcode.com/problems/two-sum/

class Solution:
    # Time complexity = O(n^2)
    def twoSum(self, nums: [int], target: int) -> [int]:
        toReturn = []
        length = len(nums)
        for i in range(length - 1):
            for j in range(i + 1, length):
                if nums[i] + nums[j] == target:
                    toReturn.append(i)
                    toReturn.append(j)
                    return toReturn


class Solution2:
    # Time Complexity = O(n)
    def twoSum(self, nums: [int], target: int) -> [int]:
        length = len(nums)
        hashMap = {nums[i]: i for i in range(length)}
        for i in range(length):
            complement = target - nums[i]
            if complement in hashMap and hashMap[complement] != i:
                return [i, hashMap[complement]]


if __name__ == "__main__":
    solution = Solution2()
    result = solution.twoSum([0, 4, 2, 0], 0)
    print(result)

File: C#/1. Two Sum.cs
public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        int length = nums.Length;
        Dictionary<int, int> map = new Dictionary<int, int>();

        for (int i = 0; i < length; i++){
            map[nums[i]] = i;
        }
        
        for (int i = 0; i < length; i++){
            int comp = target - nums[i];
            if (map.ContainsKey(comp) && map[comp] != i){
                return new int[]{i, map[comp]};
            }
        }
        
        return new int[]{-1, -1};
    }
}